proposals/0006-gep.md

```diff
@@ -0,0 +1,99 @@
+
+## Motivation
+
+Streaming of LLM responses is an important UX feature for AI chatbots. As chatbot use cases increase in complexity,
+response times from LLMs also increase. Streaming provide an enhanced UX as the end user actively sees the LLM generating the response.
+Additionally, Glide has the intention to support other streaming uses cases such as speech-to-text and text-to-speech.
+These intial efforts will support additional streaming use cases.
+
+### Requirements
+
+- R1: Streaming should be supported via the standard `v1/language` endpoint.
+- R2: Streaming should be enabled by default.
+- R3: Streaming should be specified in the Config if non-default values are needed.
+- R4: Streaming should be handled at the router level. A router is either streaming or not streaming.
+- R5: Initially, failures should be handled via a retry of the entire input prompt/message.
+- R6: Responses must be unified to the Glide Unified Response
+
+## Design
+
+First, implement a new method for each provider called `StreamChat`. For example:
+
+```go
+func (c *Client) StreamChat(ctx context.Context, request *schemas.UnifiedChatRequest) (chan *schemas.UnifiedChatResponse, error) {
+	// Create a new chat request
+	chatRequest := c.createChatRequestSchema(request)
+	// Create channels for receiving responses and errors
+	responseChannel := make(chan *schemas.UnifiedChatResponse, 100)
+	errChannel := make(chan error, 100)
+
+	fmt.Println("Starting streaming chat request")
+	
+	c.doStreamingChatRequest(ctx, chatRequest, responseChannel, errChannel)
+
+	fmt.Println("Finished streaming chat request")
+
+	// Create a channel to send individual responses
+	responseStream := make(chan *schemas.UnifiedChatResponse)
+
+	// Handle streaming responses and errors
+	go func() {
+		defer close(errChannel) // Close the channel when the function exits
+		defer close(responseStream) // Close the responseStream channel when the function exits
+		defer close(responseChannel)
+		for {
+			select {
+			case chatResponse := <-responseChannel:
+				if chatResponse != nil {
+					// Send non-nil responses to the responseStream channel
+					responseStream <- chatResponse
+				}
+			case err := <-errChannel:
+				if err != nil {
+					// Handle the error
+					fmt.Println("Received error:", err)
+				}
+			default:
+				fmt.Println("DEFAULT")
+			}
+		}
+	}()
+
+	// Return the responseStream channel
+	return responseStream, nil
+}
+```
+
+Update LangModelProvider to the following:
+
+```go
+type LangModelProvider interface {
+	Provider() string
+	Chat(ctx context.Context, request *schemas.UnifiedChatRequest) (*schemas.UnifiedChatResponse, error)
+	StreamChat(ctx context.Context, request *schemas.UnifiedChatRequest) (chan *schemas.UnifiedChatResponse, chan error)
+}
+```
+
+- The response object is a channel with unified response chunks.
```

Comment:
Should we return it as a channel or try return it as individual response objects?

Reply:
That's a good question. I would tackle this problem from the end.
We need to pick between SSE/websocket protocol for Glide API and then see what each particular connection/request handling would look like in Hertz. I'm pretty sure that would help us a lot to understand our options on the router side.

Let me know if that makes sense

Reply:
Try to make a draft with that part and let's continue that convo there to see what's the best option here

Reply:
Sounds good I'll add something about this. I'm a bit unfamiliar with both. Do you have a suggestions about which of these is best for our application?